## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#2be15dc5-5bbc-4416-b705-ae7e91431391) |
| mint-gateway-s3.sh | :x: [more...](#66c989bc-b371-475e-ba7d-85f2054be9ea) |
| mint-erasure.sh | :x: [more...](#497b0abe-94b3-4a40-ad7d-588f7b1ddf93) |
| mint-dist-erasure.sh | :x: [more...](#7c923727-597a-40b3-a91c-c77cd07ed247) |
| mint-zoned.sh | :x: [more...](#345961af-b604-4413-901d-9d53c8abb6f1) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#dcabc9f9-2b2c-4122-8982-675060982cd7) |
<a name="66c989bc-b371-475e-ba7d-85f2054be9ea"></a>
### 11951-02239c0/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30491
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5044bb4d1c05:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... FAILED in 9 seconds
{
  "name": "awscli",
  "duration": 2077,
  "function": "aws --endpoint-url http://minio-dev7.minio.io:30491 s3api put-object --bucket awscli-mint-test-bucket-28376 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (NoSuchKey) when calling the PutObject operation: The specified key does not exist."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... FAILED in 1 seconds
{
  "name": "mc",
  "duration": "179",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-13507 --quiet --no-color cp /mint/data/datafile-0-b myminio/mc-test-bucket-22319/mc-test-object-11007  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-22319/mc-test-object-11007`\nmc: <ERROR> Failed to copy `/mint/data/datafile-0-b`. Object does not exist\nTotal: 0 B, Transferred: 0 B, Speed: 0 B/s"
}
(6/15) Running minio-dotnet tests ... done in 28 seconds
(7/15) Running minio-go tests ... FAILED in 2 seconds
{
  "args": {
    "bucketName": "minio-go-test-t9t3xgmsytnxnuvu",
    "objectName": "minio-go-test-t9t3xgmsytnxnuvuunique",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "Internal": {
        "SourceVersionID": "",
        "SourceETag": "",
        "ReplicationStatus": "",
        "SourceMTime": "0001-01-01T00:00:00Z"
      }
    },
    "size": 0
  },
  "duration": 12,
  "error": "The specified key does not exist.",
  "function": "PutObject(bucketName, objectName, reader, size, opts)",
  "message": "PutObjectWithSize failed",
  "name": "minio-go: testPutObject0ByteV2",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 9 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[zero sized object]",
  "duration": 9,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchKey, message = The specified key does not exist., bucketName = minio-java-test-3942i1t, objectName = minio-java-test-1qsmr59, resource = /minio-java-test-3942i1t/minio-java-test-1qsmr59, requestId = 1671E2CAB1200F15, hostId = 6ddad6e6-0f1d-47ef-97bc-40897d5f478d)\nrequest={method=PUT, url=http://minio-dev7.minio.io:30491/minio-java-test-3942i1t/minio-java-test-1qsmr59, headers=Content-Type: application/octet-stream\nHost: minio-dev7.minio.io:30491\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210402T002501Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210402/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 371\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1671E2CAB1200F15\nX-Xss-Protection: 1; mode=block\nDate: Fri, 02 Apr 2021 00:25:01 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:853), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 34 seconds
(8/15) Running minio-py tests ... FAILED in 16 seconds
{
  "name": "minio-py:test_negative_put_object_with_path_segment",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830",
    "object_name": "/a/b/c/e6478a39-ec52-49e3-b7af-591077bb803d"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830, request_id: 1671E2D651D0F0C8, host_id: 6ddad6e6-0f1d-47ef-97bc-40897d5f478d, bucket_name: minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 738, in test_negative_put_object_with_path_segment\n    _CLIENT.put_object(bucket_name, object_name, io.BytesIO(b''), 0)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: NoSuchKey, message: The specified key does not exist., resource: /minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830//a/b/c/e6478a39-ec52-49e3-b7af-591077bb803d, request_id: 1671E2D6517672E0, host_id: 6ddad6e6-0f1d-47ef-97bc-40897d5f478d, bucket_name: minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830, object_name: a/b/c/e6478a39-ec52-49e3-b7af-591077bb803d\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 744, in test_negative_put_object_with_path_segment\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830, request_id: 1671E2D651D0F0C8, host_id: 6ddad6e6-0f1d-47ef-97bc-40897d5f478d, bucket_name: minio-py-test-d647b6b5-1b72-4e07-a585-3737ac23d830\n",
  "duration": 19
}
(8/15) Running s3cmd tests ... FAILED in 1 minutes and 1 seconds
{
  "name": "s3cmd",
  "duration": "1065",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "ERROR: S3 error: 404 (NoSuchKey): The specified key does not exist."
}
(8/15) Running s3select tests ... done in 4 seconds
(9/15) Running security tests ... done in 0 seconds

Executed 9 out of 15 tests successfully.
```
<a name="7c923727-597a-40b3-a91c-c77cd07ed247"></a>
### 11951-02239c0/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:32025
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp eba76fd5213c:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 9 seconds
{
  "name": "awscli",
  "duration": 2212,
  "function": "aws --endpoint-url http://minio-dev8.minio.io:32025 s3api put-object --bucket awscli-mint-test-bucket-4190 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (NoSuchKey) when calling the PutObject operation: The specified key does not exist."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... FAILED in 7 seconds
{
  "name": "mc",
  "duration": "211",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-25203 --quiet --no-color cp /mint/data/datafile-0-b myminio/mc-test-bucket-361/mc-test-object-10304  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-361/mc-test-object-10304`\nmc: <ERROR> Failed to copy `/mint/data/datafile-0-b`. Object does not exist\nTotal: 0 B, Transferred: 0 B, Speed: 0 B/s"
}
(6/15) Running minio-dotnet tests ... done in 38 seconds
(7/15) Running minio-go tests ... FAILED in 9 seconds
{
  "args": {
    "bucketName": "minio-go-test-llhqpnblrn4l5y1a",
    "objectName": "minio-go-test-llhqpnblrn4l5y1aunique",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "Internal": {
        "SourceVersionID": "",
        "SourceETag": "",
        "ReplicationStatus": "",
        "SourceMTime": "0001-01-01T00:00:00Z"
      }
    },
    "size": 0
  },
  "duration": 177,
  "error": "The specified key does not exist.",
  "function": "PutObject(bucketName, objectName, reader, size, opts)",
  "message": "PutObjectWithSize failed",
  "name": "minio-go: testPutObject0ByteV2",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 45 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[zero sized object]",
  "duration": 27,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchKey, message = The specified key does not exist., bucketName = minio-java-test-smj7mu, objectName = minio-java-test-1o1c8t4, resource = /minio-java-test-smj7mu/minio-java-test-1o1c8t4, requestId = 1671E2D9D8EB0C12, hostId = a5f2b772-82e4-444f-a80e-6baae677352b)\nrequest={method=PUT, url=http://minio-dev8.minio.io:32025/minio-java-test-smj7mu/minio-java-test-1o1c8t4, headers=Content-Type: application/octet-stream\nHost: minio-dev8.minio.io:32025\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210402T002606Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210402/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 369\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1671E2D9D8EB0C12\nX-Xss-Protection: 1; mode=block\nDate: Fri, 02 Apr 2021 00:26:06 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:853), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 49 seconds
(8/15) Running minio-py tests ... FAILED in 19 seconds
{
  "name": "minio-py:test_negative_put_object_with_path_segment",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2",
    "object_name": "/a/b/c/1adbaad3-16ad-46b7-b498-3574253e619f"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2, request_id: 1671E2E9BB332CFC, host_id: a5f2b772-82e4-444f-a80e-6baae677352b, bucket_name: minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 738, in test_negative_put_object_with_path_segment\n    _CLIENT.put_object(bucket_name, object_name, io.BytesIO(b''), 0)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: NoSuchKey, message: The specified key does not exist., resource: /minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2//a/b/c/1adbaad3-16ad-46b7-b498-3574253e619f, request_id: 1671E2E9B975ED3E, host_id: a5f2b772-82e4-444f-a80e-6baae677352b, bucket_name: minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2, object_name: a/b/c/1adbaad3-16ad-46b7-b498-3574253e619f\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 744, in test_negative_put_object_with_path_segment\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2, request_id: 1671E2E9BB332CFC, host_id: a5f2b772-82e4-444f-a80e-6baae677352b, bucket_name: minio-py-test-2433b91a-c733-448f-b943-c7a3d6f057d2\n",
  "duration": 111
}
(8/15) Running s3cmd tests ... FAILED in 11 seconds
{
  "name": "s3cmd",
  "duration": "1156",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "ERROR: S3 error: 404 (NoSuchKey): The specified key does not exist."
}
(8/15) Running s3select tests ... done in 8 seconds
(9/15) Running security tests ... done in 0 seconds

Executed 9 out of 15 tests successfully.
```
<a name="2be15dc5-5bbc-4416-b705-ae7e91431391"></a>
### 11951-02239c0/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31465
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9aa45f61ed48:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 9 seconds
{
  "name": "awscli",
  "duration": 2145,
  "function": "aws --endpoint-url http://minio-dev7.minio.io:31465 s3api put-object --bucket awscli-mint-test-bucket-6240 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (NoSuchKey) when calling the PutObject operation: The specified key does not exist."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... FAILED in 12 seconds
{
  "name": "mc",
  "duration": "214",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-2187 --quiet --no-color cp /mint/data/datafile-0-b myminio/mc-test-bucket-15926/mc-test-object-24834  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-15926/mc-test-object-24834`\nmc: <ERROR> Failed to copy `/mint/data/datafile-0-b`. Object does not exist\nTotal: 0 B, Transferred: 0 B, Speed: 0 B/s"
}
(6/15) Running minio-dotnet tests ... done in 41 seconds
(7/15) Running minio-go tests ... FAILED in 14 seconds
{
  "args": {
    "bucketName": "minio-go-test-4ci9ilokfxa2fjus",
    "objectName": "minio-go-test-4ci9ilokfxa2fjusunique",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "Internal": {
        "SourceVersionID": "",
        "SourceETag": "",
        "ReplicationStatus": "",
        "SourceMTime": "0001-01-01T00:00:00Z"
      }
    },
    "size": 0
  },
  "duration": 116,
  "error": "The specified key does not exist.",
  "function": "PutObject(bucketName, objectName, reader, size, opts)",
  "message": "PutObjectWithSize failed",
  "name": "minio-go: testPutObject0ByteV2",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 52 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[zero sized object]",
  "duration": 47,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchKey, message = The specified key does not exist., bucketName = minio-java-test-3c6h9gq, objectName = minio-java-test-3i8u7f9, resource = /minio-java-test-3c6h9gq/minio-java-test-3i8u7f9, requestId = 1671E2DEEBDBA153, hostId = 9eae4f5a-8b83-4329-a16d-c11ca79fea70)\nrequest={method=PUT, url=http://minio-dev7.minio.io:31465/minio-java-test-3c6h9gq/minio-java-test-3i8u7f9, headers=Content-Type: application/octet-stream\nHost: minio-dev7.minio.io:31465\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210402T002628Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210402/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 371\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1671E2DEEBDBA153\nX-Xss-Protection: 1; mode=block\nDate: Fri, 02 Apr 2021 00:26:28 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:853), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 59 seconds
(8/15) Running minio-py tests ... FAILED in 20 seconds
{
  "name": "minio-py:test_negative_put_object_with_path_segment",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-cc758457-d675-478b-8321-0e234d6850c9",
    "object_name": "/a/b/c/86b48500-efd9-40b7-af10-e4166edc6cef"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-cc758457-d675-478b-8321-0e234d6850c9, request_id: 1671E2F1506A9D5A, host_id: 9eae4f5a-8b83-4329-a16d-c11ca79fea70, bucket_name: minio-py-test-cc758457-d675-478b-8321-0e234d6850c9",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 738, in test_negative_put_object_with_path_segment\n    _CLIENT.put_object(bucket_name, object_name, io.BytesIO(b''), 0)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: NoSuchKey, message: The specified key does not exist., resource: /minio-py-test-cc758457-d675-478b-8321-0e234d6850c9//a/b/c/86b48500-efd9-40b7-af10-e4166edc6cef, request_id: 1671E2F14E9B88D6, host_id: 9eae4f5a-8b83-4329-a16d-c11ca79fea70, bucket_name: minio-py-test-cc758457-d675-478b-8321-0e234d6850c9, object_name: a/b/c/86b48500-efd9-40b7-af10-e4166edc6cef\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 744, in test_negative_put_object_with_path_segment\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-cc758457-d675-478b-8321-0e234d6850c9, request_id: 1671E2F1506A9D5A, host_id: 9eae4f5a-8b83-4329-a16d-c11ca79fea70, bucket_name: minio-py-test-cc758457-d675-478b-8321-0e234d6850c9\n",
  "duration": 177
}
(8/15) Running s3cmd tests ... FAILED in 11 seconds
{
  "name": "s3cmd",
  "duration": "1277",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "ERROR: S3 error: 404 (NoSuchKey): The specified key does not exist."
}
(8/15) Running s3select tests ... done in 8 seconds
(9/15) Running security tests ... done in 0 seconds

Executed 9 out of 15 tests successfully.
```
<a name="dcabc9f9-2b2c-4122-8982-675060982cd7"></a>
### 11951-02239c0/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32033
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8798877a7fb6:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 9 seconds
{
  "name": "awscli",
  "duration": 1886,
  "function": "aws --endpoint-url http://minio-dev4.minio.io:32033 s3api put-object --bucket awscli-mint-test-bucket-32065 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (NoSuchKey) when calling the PutObject operation: The specified key does not exist."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... FAILED in 13 seconds
{
  "name": "mc",
  "duration": "204",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-16714 --quiet --no-color cp /mint/data/datafile-0-b myminio/mc-test-bucket-5463/mc-test-object-6377  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-5463/mc-test-object-6377`\nmc: <ERROR> Failed to copy `/mint/data/datafile-0-b`. Object does not exist\nTotal: 0 B, Transferred: 0 B, Speed: 0 B/s"
}
(6/15) Running minio-dotnet tests ... done in 40 seconds
(7/15) Running minio-go tests ... FAILED in 7 seconds
{
  "args": {
    "bucketName": "minio-go-test-kzlb4t6k19yenb4f",
    "objectName": "minio-go-test-kzlb4t6k19yenb4funique",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "Internal": {
        "SourceVersionID": "",
        "SourceETag": "",
        "ReplicationStatus": "",
        "SourceMTime": "0001-01-01T00:00:00Z"
      }
    },
    "size": 0
  },
  "duration": 105,
  "error": "The specified key does not exist.",
  "function": "PutObject(bucketName, objectName, reader, size, opts)",
  "message": "PutObjectWithSize failed",
  "name": "minio-go: testPutObject0ByteV2",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 1 minutes and 12 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[zero sized object]",
  "duration": 53,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchKey, message = The specified key does not exist., bucketName = minio-java-test-33pgunh, objectName = minio-java-test-3e4tl86, resource = /minio-java-test-33pgunh/minio-java-test-3e4tl86, requestId = 1671E2E130BAEC76, hostId = ba20c26f-4fdb-48c1-9874-8c5076e53957)\nrequest={method=PUT, url=http://minio-dev4.minio.io:32033/minio-java-test-33pgunh/minio-java-test-3e4tl86, headers=Content-Type: application/octet-stream\nHost: minio-dev4.minio.io:32033\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210402T002638Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210402/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 371\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1671E2E130BAEC76\nX-Xss-Protection: 1; mode=block\nDate: Fri, 02 Apr 2021 00:26:38 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:853), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 56 seconds
(8/15) Running minio-py tests ... FAILED in 18 seconds
{
  "name": "minio-py:test_negative_put_object_with_path_segment",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5",
    "object_name": "/a/b/c/a3944130-8995-4fb6-b2d2-5934f60cfae1"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5, request_id: 1671E2F29903B3A8, host_id: ba20c26f-4fdb-48c1-9874-8c5076e53957, bucket_name: minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 738, in test_negative_put_object_with_path_segment\n    _CLIENT.put_object(bucket_name, object_name, io.BytesIO(b''), 0)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: NoSuchKey, message: The specified key does not exist., resource: /minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5//a/b/c/a3944130-8995-4fb6-b2d2-5934f60cfae1, request_id: 1671E2F297C526BD, host_id: ba20c26f-4fdb-48c1-9874-8c5076e53957, bucket_name: minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5, object_name: a/b/c/a3944130-8995-4fb6-b2d2-5934f60cfae1\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 744, in test_negative_put_object_with_path_segment\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5, request_id: 1671E2F29903B3A8, host_id: ba20c26f-4fdb-48c1-9874-8c5076e53957, bucket_name: minio-py-test-fe917791-40d8-4003-a0d8-e8a20be434e5\n",
  "duration": 92
}
(8/15) Running s3cmd tests ... FAILED in 6 seconds
{
  "name": "s3cmd",
  "duration": "3346",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-14860/s3cmd-test-object-24151 ed40f264-0ff1-4e79-9b2b-e4dcb6dca518\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id ed40f264-0ff1-4e79-9b2b-e4dcb6dca518 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(8/15) Running s3select tests ... done in 8 seconds
(9/15) Running security tests ... done in 0 seconds

Executed 9 out of 15 tests successfully.
```
<a name="497b0abe-94b3-4a40-ad7d-588f7b1ddf93"></a>
### 11951-02239c0/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:32654
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c064b328daff:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... FAILED in 9 seconds
{
  "name": "awscli",
  "duration": 1978,
  "function": "aws --endpoint-url http://minio-c2.minio.io:32654 s3api put-object --bucket awscli-mint-test-bucket-24010 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (NoSuchKey) when calling the PutObject operation: The specified key does not exist."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... FAILED in 13 seconds
{
  "name": "mc",
  "duration": "231",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-8186 --quiet --no-color cp /mint/data/datafile-0-b myminio/mc-test-bucket-10397/mc-test-object-2349  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-10397/mc-test-object-2349`\nmc: <ERROR> Failed to copy `/mint/data/datafile-0-b`. Object does not exist\nTotal: 0 B, Transferred: 0 B, Speed: 0 B/s"
}
(6/15) Running minio-dotnet tests ... done in 32 seconds
(7/15) Running minio-go tests ... FAILED in 4 seconds
{
  "args": {
    "bucketName": "minio-go-test-a0tbmbnyhcoqwwq4",
    "objectName": "minio-go-test-a0tbmbnyhcoqwwq4unique",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "Internal": {
        "SourceVersionID": "",
        "SourceETag": "",
        "ReplicationStatus": "",
        "SourceMTime": "0001-01-01T00:00:00Z"
      }
    },
    "size": 0
  },
  "duration": 64,
  "error": "The specified key does not exist.",
  "function": "PutObject(bucketName, objectName, reader, size, opts)",
  "message": "PutObjectWithSize failed",
  "name": "minio-go: testPutObject0ByteV2",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 29 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[zero sized object]",
  "duration": 20,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchKey, message = The specified key does not exist., bucketName = minio-java-test-7sjpj7, objectName = minio-java-test-3jaqk96, resource = /minio-java-test-7sjpj7/minio-java-test-3jaqk96, requestId = 1671E2D475EF7480, hostId = 8b3b9abc-afa3-4140-9f15-00e47c7cb231)\nrequest={method=PUT, url=http://minio-c2.minio.io:32654/minio-java-test-7sjpj7/minio-java-test-3jaqk96, headers=Content-Type: application/octet-stream\nHost: minio-c2.minio.io:32654\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210402T002543Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210402/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 369\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1671E2D475EF7480\nX-Xss-Protection: 1; mode=block\nDate: Fri, 02 Apr 2021 00:25:43 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:853), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 47 seconds
(8/15) Running minio-py tests ... FAILED in 17 seconds
{
  "name": "minio-py:test_negative_put_object_with_path_segment",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5",
    "object_name": "/a/b/c/1b1f2e9a-ca2f-419c-90fe-a80ec608629a"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5, request_id: 1671E2E383DB4884, host_id: 8b3b9abc-afa3-4140-9f15-00e47c7cb231, bucket_name: minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 738, in test_negative_put_object_with_path_segment\n    _CLIENT.put_object(bucket_name, object_name, io.BytesIO(b''), 0)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: NoSuchKey, message: The specified key does not exist., resource: /minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5//a/b/c/1b1f2e9a-ca2f-419c-90fe-a80ec608629a, request_id: 1671E2E382D8503A, host_id: 8b3b9abc-afa3-4140-9f15-00e47c7cb231, bucket_name: minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5, object_name: a/b/c/1b1f2e9a-ca2f-419c-90fe-a80ec608629a\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 744, in test_negative_put_object_with_path_segment\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5, request_id: 1671E2E383DB4884, host_id: 8b3b9abc-afa3-4140-9f15-00e47c7cb231, bucket_name: minio-py-test-1af31d15-5557-4ab8-ae8e-2d1703cad0a5\n",
  "duration": 69
}
(8/15) Running s3cmd tests ... FAILED in 12 seconds
{
  "name": "s3cmd",
  "duration": "1061",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "ERROR: S3 error: 404 (NoSuchKey): The specified key does not exist."
}
(8/15) Running s3select tests ... done in 3 seconds
(9/15) Running security tests ... done in 0 seconds

Executed 9 out of 15 tests successfully.
```
<a name="345961af-b604-4413-901d-9d53c8abb6f1"></a>
### 11951-02239c0/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:32604
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 537cfa6c80c2:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 9 seconds
{
  "name": "awscli",
  "duration": 2292,
  "function": "aws --endpoint-url http://minio-dev8.minio.io:32604 s3api put-object --bucket awscli-mint-test-bucket-17435 --key prefix/directory/\n",
  "status": "FAIL",
  "error": "An error occurred (NoSuchKey) when calling the PutObject operation: The specified key does not exist."
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... FAILED in 7 seconds
{
  "name": "mc",
  "duration": "204",
  "function": "test_put_object_0byte",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-16249 --quiet --no-color cp /mint/data/datafile-0-b myminio/mc-test-bucket-24901/mc-test-object-1682  >>> \n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-24901/mc-test-object-1682`\nmc: <ERROR> Failed to copy `/mint/data/datafile-0-b`. Object does not exist\nTotal: 0 B, Transferred: 0 B, Speed: 0 B/s"
}
(6/15) Running minio-dotnet tests ... done in 40 seconds
(7/15) Running minio-go tests ... FAILED in 9 seconds
{
  "args": {
    "bucketName": "minio-go-test-q35jjw9wd9e1cdrl",
    "objectName": "minio-go-test-q35jjw9wd9e1cdrlunique",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "Internal": {
        "SourceVersionID": "",
        "SourceETag": "",
        "ReplicationStatus": "",
        "SourceMTime": "0001-01-01T00:00:00Z"
      }
    },
    "size": 0
  },
  "duration": 98,
  "error": "The specified key does not exist.",
  "function": "PutObject(bucketName, objectName, reader, size, opts)",
  "message": "PutObjectWithSize failed",
  "name": "minio-go: testPutObject0ByteV2",
  "status": "FAIL"
}
(7/15) Running minio-java tests ... FAILED in 59 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[zero sized object]",
  "duration": 46,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchKey, message = The specified key does not exist., bucketName = minio-java-test-1l4li5n, objectName = minio-java-test-33mo3o1, resource = /minio-java-test-1l4li5n/minio-java-test-33mo3o1, requestId = 1671E2DDE030C0E8, hostId = dab00cdf-c1a6-4431-bee6-e67e17e26eab)\nrequest={method=PUT, url=http://minio-dev8.minio.io:32604/minio-java-test-1l4li5n/minio-java-test-33mo3o1, headers=Content-Type: application/octet-stream\nHost: minio-dev8.minio.io:32604\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210402T002623Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210402/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 371\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1671E2DDE030C0E8\nX-Xss-Protection: 1; mode=block\nDate: Fri, 02 Apr 2021 00:26:23 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.putObject(MinioClient.java:4547), io.minio.MinioClient.putObject(MinioClient.java:2713), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:853), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(7/15) Running minio-js tests ... done in 1 minutes and 2 seconds
(8/15) Running minio-py tests ... FAILED in 21 seconds
{
  "name": "minio-py:test_negative_put_object_with_path_segment",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6",
    "object_name": "/a/b/c/18adcd83-f993-4fef-a4c8-64f55a173f42"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6, request_id: 1671E2F1065C57BC, host_id: dab00cdf-c1a6-4431-bee6-e67e17e26eab, bucket_name: minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 738, in test_negative_put_object_with_path_segment\n    _CLIENT.put_object(bucket_name, object_name, io.BytesIO(b''), 0)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1747, in put_object\n    bucket_name, object_name, part_data, headers,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: NoSuchKey, message: The specified key does not exist., resource: /minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6//a/b/c/18adcd83-f993-4fef-a4c8-64f55a173f42, request_id: 1671E2F103DAEF77, host_id: dab00cdf-c1a6-4431-bee6-e67e17e26eab, bucket_name: minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6, object_name: a/b/c/18adcd83-f993-4fef-a4c8-64f55a173f42\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 744, in test_negative_put_object_with_path_segment\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6, request_id: 1671E2F1065C57BC, host_id: dab00cdf-c1a6-4431-bee6-e67e17e26eab, bucket_name: minio-py-test-a1652b83-7036-49c2-8885-1474bb9dbda6\n",
  "duration": 150
}
(8/15) Running s3cmd tests ... FAILED in 11 seconds
{
  "name": "s3cmd",
  "duration": "1176",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "ERROR: S3 error: 404 (NoSuchKey): The specified key does not exist."
}
(8/15) Running s3select tests ... done in 8 seconds
(9/15) Running security tests ... done in 0 seconds

Executed 9 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -12,7 +12,7 @@ jobs:
     strategy:
       matrix:
         go-version: [1.16.x]
-        os: [ubuntu-latest, windows-latest, macos-latest]
+        os: [ubuntu-latest, windows-latest]
     steps:
       - uses: actions/checkout@v2
       - uses: actions/setup-node@v1
@@ -21,14 +21,6 @@ jobs:
       - uses: actions/setup-go@v2
         with:
           go-version: ${{ matrix.go-version }}
-      - name: Build on ${{ matrix.os }}
-        if: matrix.os == 'macos-latest'
-        env:
-          CGO_ENABLED: 0
-          GO111MODULE: on
-        run: |
-          make
-          make test-race
       - name: Build on ${{ matrix.os }}
         if: matrix.os == 'windows-latest'
         env:

--- a/cmd/erasure-healing.go
+++ b/cmd/erasure-healing.go
@@ -376,9 +376,14 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 		partsMetadata[i] = cleanFileInfo(latestMeta)
 	}
 
-	dataDir := latestMeta.DataDir
+	// source data dir shall be empty in case of XLV1
+	// differentiate it with dstDataDir for readability
+	// srcDataDir is the one used with newBitrotReader()
+	// to read existing content.
+	srcDataDir := latestMeta.DataDir
+	dstDataDir := latestMeta.DataDir
 	if latestMeta.XLV1 {
-		dataDir = migrateDataDir
+		dstDataDir = migrateDataDir
 	}
 
 	var inlineBuffers []*bytes.Buffer
@@ -419,10 +424,7 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 					continue
 				}
 				checksumInfo := copyPartsMetadata[i].Erasure.GetChecksumInfo(partNumber)
-				partPath := pathJoin(object, dataDir, fmt.Sprintf("part.%d", partNumber))
-				if latestMeta.XLV1 {
-					partPath = pathJoin(object, fmt.Sprintf("part.%d", partNumber))
-				}
+				partPath := pathJoin(object, srcDataDir, fmt.Sprintf("part.%d", partNumber))
 				readers[i] = newBitrotReader(disk, partsMetadata[i].Data, bucket, partPath, tillOffset, checksumAlgo, checksumInfo.Hash, erasure.ShardSize())
 			}
 			writers := make([]io.Writer, len(outDatedDisks))
@@ -430,7 +432,7 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 				if disk == OfflineDisk {
 					continue
 				}
-				partPath := pathJoin(tmpID, dataDir, fmt.Sprintf("part.%d", partNumber))
+				partPath := pathJoin(tmpID, dstDataDir, fmt.Sprintf("part.%d", partNumber))
 				if len(inlineBuffers) > 0 {
 					inlineBuffers[i] = bytes.NewBuffer(make([]byte, 0, erasure.ShardFileSize(latestMeta.Size)))
 					writers[i] = newStreamingBitrotWriterBuffer(inlineBuffers[i], DefaultBitrotAlgorithm, erasure.ShardSize())
@@ -460,7 +462,7 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 					continue
 				}
 
-				partsMetadata[i].DataDir = dataDir
+				partsMetadata[i].DataDir = dstDataDir
 				partsMetadata[i].AddObjectPart(partNumber, "", partSize, partActualSize)
 				partsMetadata[i].Erasure.AddChecksumInfo(ChecksumInfo{
 					PartNumber: partNumber,
@@ -481,25 +483,6 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 		}
 	}
 
-	if len(inlineBuffers) > 0 {
-		// Write directly...
-		if outDatedDisks, err = writeUniqueFileInfo(ctx, outDatedDisks, bucket, object, partsMetadata, diskCount(outDatedDisks)); err != nil {
-			logger.LogIf(ctx, err)
-			return result, toObjectErr(err, bucket, object)
-		}
-		result.ObjectSize = latestMeta.Size
-		for _, disk := range outDatedDisks {
-			if disk == OfflineDisk {
-				continue
-			}
-			for i, v := range result.Before.Drives {
-				if v.Endpoint == disk.String() {
-					result.After.Drives[i].State = madmin.DriveStateOk
-				}
-			}
-		}
-		return result, nil
-	}
 	defer er.deleteObject(context.Background(), minioMetaTmpBucket, tmpID, len(storageDisks)/2+1)
 
 	// Generate and write `xl.meta` generated from other disks.

--- a/cmd/erasure-metadata.go
+++ b/cmd/erasure-metadata.go
@@ -279,6 +279,37 @@ func pickValidFileInfo(ctx context.Context, metaArr []FileInfo, modTime time.Tim
 	return findFileInfoInQuorum(ctx, metaArr, modTime, quorum)
 }
 
+// Rename metadata content to destination location for each disk concurrently.
+func renameFileInfo(ctx context.Context, disks []StorageAPI, srcBucket, srcEntry, dstBucket, dstEntry string, quorum int) ([]StorageAPI, error) {
+	ignoredErr := []error{errFileNotFound}
+
+	g := errgroup.WithNErrs(len(disks))
+
+	// Rename file on all underlying storage disks.
+	for index := range disks {
+		index := index
+		g.Go(func() error {
+			if disks[index] == nil {
+				return errDiskNotFound
+			}
+			if err := disks[index].RenameData(ctx, srcBucket, srcEntry, "", dstBucket, dstEntry); err != nil {
+				if !IsErrIgnored(err, ignoredErr...) {
+					return err
+				}
+			}
+			return nil
+		}, index)
+	}
+
+	// Wait for all renames to finish.
+	errs := g.Wait()
+
+	// We can safely allow RenameData errors up to len(er.getDisks()) - writeQuorum
+	// otherwise return failure. Cleanup successful renames.
+	err := reduceWriteQuorumErrs(ctx, errs, objectOpIgnoredErrs, quorum)
+	return evalDisks(disks, errs), err
+}
+
 // writeUniqueFileInfo - writes unique `xl.meta` content for each disk concurrently.
 func writeUniqueFileInfo(ctx context.Context, disks []StorageAPI, bucket, prefix string, files []FileInfo, quorum int) ([]StorageAPI, error) {
 	g := errgroup.WithNErrs(len(disks))

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -124,11 +124,27 @@ func (er erasureObjects) CopyObject(ctx context.Context, srcBucket, srcObject, d
 		}
 	}
 
-	// Write directly...
-	if _, err = writeUniqueFileInfo(ctx, onlineDisks, srcBucket, srcObject, metaArr, writeQuorum); err != nil {
-		logger.LogIf(ctx, err)
-		return ObjectInfo{}, toObjectErr(err, srcBucket, srcObject)
+	tempObj := mustGetUUID()
+
+	var online int
+	// Cleanup in case of xl.meta writing failure
+	defer func() {
+		if online != len(onlineDisks) {
+			er.deleteObject(context.Background(), minioMetaTmpBucket, tempObj, writeQuorum)
+		}
+	}()
+
+	// Write unique `xl.meta` for each disk.
+	if onlineDisks, err = writeUniqueFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, metaArr, writeQuorum); err != nil {
+		return oi, toObjectErr(err, srcBucket, srcObject)
 	}
+
+	// Rename atomically `xl.meta` from tmp location to destination for each disk.
+	if _, err = renameFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, srcBucket, srcObject, writeQuorum); err != nil {
+		return oi, toObjectErr(err, srcBucket, srcObject)
+	}
+
+	online = countOnlineDisks(onlineDisks)
 
 	return fi.ToObjectInfo(srcBucket, srcObject), nil
 }
@@ -806,23 +822,15 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 	}
 
 	// Write unique `xl.meta` for each disk.
-	if len(inlineBuffers) == 0 {
-		if onlineDisks, err = writeUniqueFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, partsMetadata, writeQuorum); err != nil {
-			logger.LogIf(ctx, err)
-			return ObjectInfo{}, toObjectErr(err, bucket, object)
-		}
+	if onlineDisks, err = writeUniqueFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, partsMetadata, writeQuorum); err != nil {
+		logger.LogIf(ctx, err)
+		return ObjectInfo{}, toObjectErr(err, bucket, object)
+	}
 
-		// Rename the successfully written temporary object to final location.
-		if onlineDisks, err = renameData(ctx, onlineDisks, minioMetaTmpBucket, tempObj, fi.DataDir, bucket, object, writeQuorum, nil); err != nil {
-			logger.LogIf(ctx, err)
-			return ObjectInfo{}, toObjectErr(err, bucket, object)
-		}
-	} else {
-		// Write directly...
-		if onlineDisks, err = writeUniqueFileInfo(ctx, onlineDisks, bucket, object, partsMetadata, writeQuorum); err != nil {
-			logger.LogIf(ctx, err)
-			return ObjectInfo{}, toObjectErr(err, bucket, object)
-		}
+	// Rename the successfully written temporary object to final location.
+	if onlineDisks, err = renameData(ctx, onlineDisks, minioMetaTmpBucket, tempObj, fi.DataDir, bucket, object, writeQuorum, nil); err != nil {
+		logger.LogIf(ctx, err)
+		return ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
 
 	// Whether a disk was initially or becomes offline
@@ -1245,11 +1253,27 @@ func (er erasureObjects) PutObjectTags(ctx context.Context, bucket, object strin
 		}
 	}
 
-	// Write directly...
-	if _, err = writeUniqueFileInfo(ctx, onlineDisks, bucket, object, metaArr, writeQuorum); err != nil {
-		logger.LogIf(ctx, err)
+	tempObj := mustGetUUID()
+
+	var online int
+	// Cleanup in case of xl.meta writing failure
+	defer func() {
+		if online != len(onlineDisks) {
+			er.deleteObject(context.Background(), minioMetaTmpBucket, tempObj, writeQuorum)
+		}
+	}()
+
+	// Write unique `xl.meta` for each disk.
+	if onlineDisks, err = writeUniqueFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, metaArr, writeQuorum); err != nil {
 		return ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
+
+	// Atomically rename metadata from tmp location to destination for each disk.
+	if onlineDisks, err = renameFileInfo(ctx, onlineDisks, minioMetaTmpBucket, tempObj, bucket, object, writeQuorum); err != nil {
+		return ObjectInfo{}, toObjectErr(err, bucket, object)
+	}
+
+	online = countOnlineDisks(onlineDisks)
 
 	objInfo := fi.ToObjectInfo(bucket, object)
 	objInfo.UserTags = tags
@@ -1302,12 +1326,28 @@ func (er erasureObjects) updateObjectMeta(ctx context.Context, bucket, object st
 		metaArr[i].Metadata = fi.Metadata
 	}
 
-	// Write directly...
-	if _, err = writeUniqueFileInfo(ctx, disks, bucket, object, metaArr, writeQuorum); err != nil {
+	tempObj := mustGetUUID()
+
+	var online int
+	// Cleanup in case of xl.meta writing failure
+	defer func() {
+		if online != len(disks) {
+			er.deleteObject(context.Background(), minioMetaTmpBucket, tempObj, writeQuorum)
+		}
+	}()
+
+	// Write unique `xl.meta` for each disk.
+	if disks, err = writeUniqueFileInfo(ctx, disks, minioMetaTmpBucket, tempObj, metaArr, writeQuorum); err != nil {
+		return toObjectErr(err, bucket, object)
+	}
+
+	// Atomically rename metadata from tmp location to destination for each disk.
+	if disks, err = renameFileInfo(ctx, disks, minioMetaTmpBucket, tempObj, bucket, object, writeQuorum); err != nil {
 		logger.LogIf(ctx, err)
 		return toObjectErr(err, bucket, object)
 	}
 
+	online = countOnlineDisks(disks)
 	return nil
 }
 

--- a/cmd/xl-storage-disk-id-check.go
+++ b/cmd/xl-storage-disk-id-check.go
@@ -64,11 +64,14 @@ const (
 
 // Detects change in underlying disk.
 type xlStorageDiskIDCheck struct {
-	storage StorageAPI
-	diskID  string
-
-	apiCalls     [storageMetricLast]uint64
+	// fields position optimized for memory please
+	// do not re-order them, if you add new fields
+	// please use `fieldalignment ./...` to check
+	// if your changes are not causing any problems.
+	storage      StorageAPI
 	apiLatencies [storageMetricLast]ewma.MovingAverage
+	diskID       string
+	apiCalls     [storageMetricLast]uint64
 }
 
 func (p *xlStorageDiskIDCheck) getMetrics() DiskMetrics {

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -1613,9 +1613,6 @@ func (s *xlStorage) CheckParts(ctx context.Context, volume string, path string, 
 
 	for _, part := range fi.Parts {
 		partPath := pathJoin(path, fi.DataDir, fmt.Sprintf("part.%d", part.Number))
-		if fi.XLV1 {
-			partPath = pathJoin(path, fmt.Sprintf("part.%d", part.Number))
-		}
 		filePath := pathJoin(volumeDir, partPath)
 		if err = checkPathLength(filePath); err != nil {
 			return err
@@ -1778,10 +1775,6 @@ func (s *xlStorage) Delete(ctx context.Context, volume string, path string, recu
 
 // RenameData - rename source path to destination path atomically, metadata and data directory.
 func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath, dataDir, dstVolume, dstPath string) (err error) {
-	if dataDir == "" {
-		return errInvalidArgument
-	}
-
 	srcVolumeDir, err := s.getVolDir(srcVolume)
 	if err != nil {
 		return err
@@ -1814,11 +1807,15 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath, dataDir,
 	srcFilePath := pathutil.Join(srcVolumeDir, pathJoin(srcPath, xlStorageFormatFile))
 	dstFilePath := pathutil.Join(dstVolumeDir, pathJoin(dstPath, xlStorageFormatFile))
 
-	srcDataPath := retainSlash(pathJoin(srcVolumeDir, srcPath, dataDir))
-	// make sure to always use path.Join here, do not use pathJoin as
-	// it would additionally add `/` at the end and it comes in the
-	// way of renameAll(), parentDir creation.
-	dstDataPath := pathutil.Join(dstVolumeDir, dstPath, dataDir)
+	var srcDataPath string
+	var dstDataPath string
+	if dataDir != "" {
+		srcDataPath = retainSlash(pathJoin(srcVolumeDir, srcPath, dataDir))
+		// make sure to always use path.Join here, do not use pathJoin as
+		// it would additionally add `/` at the end and it comes in the
+		// way of renameAll(), parentDir creation.
+		dstDataPath = pathutil.Join(dstVolumeDir, dstPath, dataDir)
+	}
 
 	if err = checkPathLength(srcFilePath); err != nil {
 		return err
@@ -1994,22 +1991,33 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath, dataDir,
 		return errFileCorrupt
 	}
 
-	if err = s.WriteAll(ctx, srcVolume, pathJoin(srcPath, xlStorageFormatFile), dstBuf); err != nil {
-		return err
-	}
+	if srcDataPath != "" {
+		if err = s.WriteAll(ctx, srcVolume, pathJoin(srcPath, xlStorageFormatFile), dstBuf); err != nil {
+			return err
+		}
 
-	tmpuuid := mustGetUUID()
-	renameAll(oldDstDataPath, pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, tmpuuid))
-	tmpuuid = mustGetUUID()
-	renameAll(dstDataPath, pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, tmpuuid))
-	if err = renameAll(srcDataPath, dstDataPath); err != nil {
-		logger.LogIf(ctx, err)
-		return osErrToFileErr(err)
-	}
+		if oldDstDataPath != "" {
+			renameAll(oldDstDataPath, pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, mustGetUUID()))
+		}
+		renameAll(dstDataPath, pathutil.Join(s.diskPath, minioMetaTmpDeletedBucket, mustGetUUID()))
 
-	// Commit meta-file
-	if err = renameAll(srcFilePath, dstFilePath); err != nil {
-		return osErrToFileErr(err)
+		// renameAll only for objects that have xl.meta not saved inline.
+		if len(fi.Data) == 0 && fi.Size > 0 {
+			if err = renameAll(srcDataPath, dstDataPath); err != nil {
+				logger.LogIf(ctx, err)
+				return osErrToFileErr(err)
+			}
+		}
+
+		// Commit meta-file
+		if err = renameAll(srcFilePath, dstFilePath); err != nil {
+			return osErrToFileErr(err)
+		}
+	} else {
+		// Write meta-file directly, no data
+		if err = s.WriteAll(ctx, dstVolume, pathJoin(dstPath, xlStorageFormatFile), dstBuf); err != nil {
+			return err
+		}
 	}
 
 	// Remove parent dir of the source file if empty
@@ -2137,9 +2145,6 @@ func (s *xlStorage) VerifyFile(ctx context.Context, volume, path string, fi File
 	for _, part := range fi.Parts {
 		checksumInfo := erasure.GetChecksumInfo(part.Number)
 		partPath := pathJoin(volumeDir, path, fi.DataDir, fmt.Sprintf("part.%d", part.Number))
-		if fi.XLV1 {
-			partPath = pathJoin(volumeDir, path, fmt.Sprintf("part.%d", part.Number))
-		}
 		if err := s.bitrotVerify(partPath,
 			erasure.ShardFileSize(part.Size),
 			checksumInfo.Algorithm,

